daml-lf/tests/MultiKeys.daml

```diff
@@ -0,0 +1,232 @@
+  where
+    message "E"
+
+template Keyed
+  with
+    p : Party
+  where
+    signatory p
+    key p : Party
+    maintainer key
+
+-- All these tests operate under the assumption that there
+-- exists two global contracts with the same key `p` and one
+-- of those will always be returned by fetch/lookp-by-key at
+-- the beginning of the choice.
+-- All choices here are accepted in non-uck mode.
+-- Behavior in uck-mode is documented in comments.
+template KeyOperations
+  with
+    p : Party
+  where
+    signatory p
+
+    -- should be rejected in uck
+    nonconsuming choice CreateOverwritesLocal : ()
+      controller p
+      do cid1 <- create (Keyed p)
+         cid2 <- create (Keyed p)
+         (cid, _) <- fetchByKey @Keyed p
+         cid2 === cid
+
+    -- should be rejected in uck
+    nonconsuming choice CreateOverwritesUnknownGlobal : ()
+      controller p
+      do -- assume the global exists but do not fetch it
+         cid1 <- create (Keyed p)
+         (cid, _) <- fetchByKey @Keyed p
+         cid1 === cid
+
+    -- should be rejected in uck
+    nonconsuming choice CreateOverwritesKnownGlobal : ()
+      controller p
+      do _ <- fetchByKey @Keyed p
+         cid1 <- create (Keyed p)
+         (cid, _) <- fetchByKey @Keyed p
+         cid1 === cid
+
+    -- should be accepted in uck, cid not fetched by key
+    nonconsuming choice FetchDoesNotOverwriteGlobal : ()
+      with
+        cid : ContractId Keyed
+      controller p
+      do (cidByKey, _) <- fetchByKey @Keyed p
+         cid =/= cidByKey
+         c <- fetch cid
+         key c === p
+         (cidByKey', _) <- fetchByKey @Keyed p
+         cidByKey' === cidByKey
+
+    -- should be accepted in uck, cid not fetched by key
+    nonconsuming choice FetchDoesNotOverwriteLocal : ()
+      with
+        cid : ContractId Keyed
+      controller p
+      do local <- create (Keyed p)
+         c <- fetch cid
+         key c === p
+         (cid', _) <- fetchByKey @Keyed p
+         local === cid'
+
+    -- should be accepted in uck
+    nonconsuming choice LocalArchiveOverwritesUnknownGlobal : ()
+      controller p
+      do -- assume that a global contract with same key exists but don’t fetch it.
+         local <- create (Keyed p)
+         archive local
+         None <- lookupByKey @Keyed p
+         pure ()
+
+    -- should be rejected in uck
+    nonconsuming choice LocalArchiveOverwritesKnownGlobal : ()
+      controller p
+      do _ <- fetchByKey @Keyed p
+         local <- create (Keyed p)
+         archive local
+         None <- lookupByKey @Keyed p
+         pure ()
+
+    -- Should be accepted in uck
+    nonconsuming choice GlobalArchiveOverwritesUnknownGlobal : ()
+      with
+        cid1 : ContractId Keyed
+        cid2 : ContractId Keyed
+      controller p
+      do cid1 =/= cid2
+         c1 <- fetch cid1
+         key c1 === p
+         c2 <- fetch cid2
+         key c2 === p
+         archive cid1
+         None <- lookupByKey @Keyed p
```

Comment:
I wonder why whether `None` is required here. I'd argue the following:
1. Daml engine should ask the ledger `ResultNeedKey` at this point whether there is a contract for key `@Keyed p`.
2. If the ledger responds with `cid1`, then the answer should be `None` because `cid1` has been archived before in the transaction.
3. If the ledger responds with `cid2`, then the answer should be `Some cid2` because `cid2` has not been archived yet.

Reply:
Yes, that's my mental model that I settled upon.

Reply:
As discussed, we’re postponing multikey semantics so resolving this comment for this PR.